core/src/main/java/com/linecorp/armeria/server/docs/DocService.java

```diff
@@ -259,13 +228,129 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
         return unwrap().serve(ctx, req);
     }
 
+    static final class SpecificationLoader {
+        private final ExampleSupport exampleSupport;
+        private final DocServiceFilter filter;
+        private final NamedTypeInfoProvider namedTypeInfoProvider;
+        private final AtomicReference<List<ServiceConfig>> servicesRef =
+                new AtomicReference<>(Collections.emptyList());
+        private final Map<String, CompletableFuture<AggregatedHttpFile>> files = new ConcurrentHashMap<>();
+        private static final String versionsPath = "/versions.json";
+        private static final String specificationPath = "/specification.json";
+        private static final Set<String> targetPaths = ImmutableSet.of(versionsPath, specificationPath);
+
+        SpecificationLoader(
+                ExampleSupport exampleSupport,
+                DocServiceFilter filter,
+                @Nullable NamedTypeInfoProvider namedTypeInfoProvider) {
+            this.exampleSupport = exampleSupport;
+            this.filter = filter;
+            this.namedTypeInfoProvider = composeNamedTypeInfoProvider(namedTypeInfoProvider);
+        }
+
+        boolean supports(String path) {
+            return targetPaths.contains(path);
+        }
+
+        void triggerPreload(Executor executor) {
+            targetPaths.forEach(path -> load(path, executor));
+        }
+
+        CompletableFuture<AggregatedHttpFile> load(String path, Executor executor) {
+            if (versionsPath.equals(path)) {
+                return loadVersions(executor);
+            } else if (specificationPath.equals(path)) {
+                return loadSpecifications(executor);
+            } else {
+                throw new IllegalStateException("Unexpected path: " + path);
+            }
+        }
+
+        private CompletableFuture<AggregatedHttpFile> loadVersions(Executor executor) {
+            return files.computeIfAbsent(versionsPath, key -> CompletableFuture.supplyAsync(() -> {
+                final List<Version> versions = ImmutableList.copyOf(
+                        Version.getAll(DocService.class.getClassLoader()).values());
+                try {
+                    final byte[] content = jsonMapper.writerWithDefaultPrettyPrinter()
+                                                     .writeValueAsBytes(versions);
+                    return toFile(content, MediaType.JSON_UTF_8);
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                }
+            }, executor));
+        }
+
+        private CompletableFuture<AggregatedHttpFile> loadSpecifications(Executor executor) {
+            return files.computeIfAbsent(specificationPath, key -> {
```

Comment:
Note: By definition, checked that at least `DocService` will be added.
It's also worth noting that there isn't any logic to stop users from adding virtual hosts without any services, but I guess this in this case `DocService` won't be added either so it doesn't matter.

Reply:
Instead of doing the check you mentioned, I decided to add an explicit `get` method which doesn't attempt to `compute` the services.
The reasoning for this is to avoid throwing an exception if a server is reconfigured with an empty `DocService`.

```java
sb.reconfigure(sb -> sb.serviceUnder("/docs", DocService.of())) // services would be empty
```